doc/sphinx/proof-engine/tactics.rst

```diff
@@ -466,52 +466,87 @@ Examples:
-that ``|-`` is optional in the first case when no ``*`` is given.
-
-Here are some tactics that understand occurrence clauses: :tacn:`set`,
-:tacn:`remember`, :tacn:`induction`, :tacn:`destruct`.
+   .. insertprodn occurrences concl_occs
+
+   .. prodn::
+      occurrences ::= at @occs_nums
+      | in @goal_occurrences
+      occs_nums ::= {? - } {+ {| @natural | @ident } }
+      goal_occurrences ::= {*, @hyp_occs } {? %|- {? @concl_occs } }
+      | * %|- {? @concl_occs }
+      | * {? at @occs_nums }
+      hyp_occs ::= @hypident {? at @occs_nums }
+      hypident ::= @ident
+      | ( type of @ident )
+      | ( value of @ident )
+      concl_occs ::= * {? at @occs_nums }
+
+   :n:`@occurrences`
+     The first form of :token:`occurrences` selects occurrences in
+     the conclusion of the goal.  The second form can select occurrences
+     in the goal conclusion and in one or more hypotheses.
+
+   :n:`{? - } {+ {| @natural | @ident } }`
+     Selects the specified occurrences within a single goal or hypothesis.
+     Occurrences are numbered from left to right starting with 1 when the
+     goal is printed with the :flag:`Printing All`.  (In particular, occurrences
+     in implicit arguments (see :ref:`ImplicitArguments`) and
+     coercions (see :ref:`Coercions`) are counted but not shown by default.)
+
+     Specifying `-` includes all occurrences *except* the ones listed.
+
+   :n:`{*, @hyp_occs } {? %|- {? @concl_occs } }`
+     Selects occurrences in the specified hypotheses and the
+     specified occurrences in the conclusion.
+
+     .. todo PR: if above is empty it's useless, right?
+
+   :n:`* %|- {? @concl_occs }`
+     Selects all occurrences in all hypotheses and the
+     specified occurrences in the conclusion.
+
+   :n:`@goal_occurrences ::= * {? at @occs_nums }`
+     Selects all occurrences in all hypotheses and in the conclusion.
+
+     .. todo PR: above is likely wrong.  If * selects everything hyp+concl, the
+        occs_nums don't make sense
+
+   :n:`@hypident {? at @occs_nums }`
+     Omiting :token:`occs_nums` selects all occurrences within the hypothesis.
+
+   :n:`@hypident ::= @ident`
+     Selects the hypothesis named :token:`ident`.
+
+   :n:`( type of @ident )`
+     Selects the type part of the named hypothesis (e.g. `: nat`).
+
+   :n:`( value of @ident )`
+     Selects the value part of the named hypothesis (e.g. `:= 1`).
+
+   :n:`@concl_occs ::= * {? at @occs_nums }`
+     Selects occurrences in the conclusion.  Omitting :token:`occs_nums`
+     selects all occurrences within the conclusion.
+
+   Use `in *`` to select all occurrences in all hypotheses and the conclusion,
+   which is equivalent to `in * |- *`.  Use `* |-` to select all occurrences
+   in all hypotheses.
+
+.. todo PR: is the following true?  If so, quite a wart.
+
+   As an exception to the left-to-right order, the occurrences in
+   the return subexpression of a match are considered *before* the
+   occurrences in the matched term.
```

Comment:
It doesn't seem to be true according to this test:

```coq
Parameter b : nat.
Definition a := b.
Goal match a return let _ := a in Prop with 0 => let _ := a in True | _ => False end.
(*
H : a + a = 0
________________
a = a /\ a = 0
*)
change a with b at 2.
```

Reply:
Thanks, gone now.